src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs

```diff
@@ -22,47 +31,131 @@ namespace Microsoft.CodeAnalysis.Editor.Implementation.Diagnostics
         protected internal abstract bool IncludeDiagnostic(DiagnosticData data);
         protected internal abstract ITagSpan<TTag> CreateTagSpan(bool isLiveUpdate, SnapshotSpan span, DiagnosticData data);
 
-        ITagger<T> ITaggerProvider.CreateTagger<T>(ITextBuffer buffer)
+        protected override Task ProduceTagsAsync(TaggerContext<TTag> context, DocumentSnapshotSpan spanToTag, int? caretPosition)
         {
-            return CreateTagger<T>(buffer);
+            ProduceTags(context, spanToTag);
+            return SpecializedTasks.EmptyTask;
         }
 
-        public IAccurateTagger<T> CreateTagger<T>(ITextBuffer buffer) where T : ITag
+        private void ProduceTags(TaggerContext<TTag> context, DocumentSnapshotSpan spanToTag)
         {
-            var tagger = buffer.Properties.GetOrCreateSingletonProperty(
-                _uniqueKey, () => new AggregatingTagger(this, buffer));
-            tagger.OnTaggerCreated();
-            return tagger as IAccurateTagger<T>;
+            try
+            {
+                if (!this.IsEnabled)
+                {
+                    return;
+                }
+
+                var document = spanToTag.Document;
+                if (document == null)
+                {
+                    return;
+                }
+
+                if (!document.TryGetText(out var sourceText))
+                {
+                    return;
+                }
+                
+                var editorSnapshot = sourceText.FindCorrespondingEditorTextSnapshot();
+                if (editorSnapshot == null)
+                {
+                    return;
+                }
+
+                var cancellationToken = context.CancellationToken;
+                var workspace = document.Project.Solution.Workspace;
+                
+                // See if we've marked any spans as those we want to suppress diagnostics for.
+                // This can happen for buffers used in the preview workspace where some feature
+                // is generating code that it doesn't want errors shown for.
+                var buffer = spanToTag.SnapshotSpan.Snapshot.TextBuffer;
+                var suppressedDiagnosticsSpans = default(NormalizedSnapshotSpanCollection);
+                buffer?.Properties.TryGetProperty(PredefinedPreviewTaggerKeys.SuppressDiagnosticsSpansKey, out suppressedDiagnosticsSpans);
+
+                var eventArgs = _diagnosticService.GetDiagnosticsUpdatedEventArgs(
+                    workspace, document.Project.Id, document.Id, context.CancellationToken);
+
+                foreach (var updateArg in eventArgs)
+                {
+                    ProduceTags(
+                        context, spanToTag, workspace, document, sourceText, editorSnapshot,
+                        suppressedDiagnosticsSpans, updateArg, cancellationToken);
+                }
+            }
+            catch (ArgumentOutOfRangeException ex) when (FatalError.ReportWithoutCrash(ex))
```

Comment:
can we move this try/catch down to ProduceTags for one specific updateArg so that even if 1 fails, we still can get other analyzers squiggles?

Reply:
Sure.

Reply:
Done.